### PR TITLE
Remove use of rand r

### LINKED ISSUE #674: Don't use rand_r
rand_r doesn't seem to be available readily with cygwin, and it's not a good solution anyway. It's also not C99. We should replace it by one of the BOOST random number generators.

Occurrences:
```
source/base/utilities.cc:390:    // context. thus, use rand_r. this does not produce reproducible results
source/base/utilities.cc:396:    const double y = 1.0*rand_r(&seed)/RAND_MAX;
```

### PR DESCRIPTION
**Pull request by bangerth**
This closes #674 by implementing the solution outlined there.

### CODE REVIEW

File: source/base/utilities.cc [hunk 1]
```diff
@@ -390,41 +391,13 @@ namespace Utilities
...
-        Assert (iteration < 20, ExcInternalError());
-      };
-    return x+a;
+    // we would want to use rand(), but that function is not reentrant
+    // in a thread context. one could use rand_r, but this does not
+    // produce reproducible results between threads either (though at
+    // least it is reentrant). these two approaches being
+    // non-workable, use a thread-local random number generator here
+    static Threads::ThreadLocalStorage<boost::mt19937> random_number_generator;
+    return boost::normal_distribution<>(a,sigma)(random_number_generator.get());
```
Comment (tamiko) [inline source/base/utilities.cc]: ```normal_distribution```? if we want to emulate ```rand```, or ```rand_r```, then ```uniform_real_distribution``` it is?!

Comment (bangerth) [inline source/base/utilities.cc]: But the function returns a normally distributed number. The last line here replaces the 30 lines of Newton method :-)

Comment (tamiko) [inline source/base/utilities.cc]: > But the function returns a normally distributed number. The last line
> here replaces the 30 lines of Newton method :-)

Ah, ic. How boring ;-)

**Comment by tamiko** [comment]
Do you mind to remove the macro <code>DEAL_II_HAVE_RAND_R</code> in <code>config.h.in</code> and the macro name at the top, as well as the check in <code>cmake/checks/check_02_system_features.cmake</code>?

**Comment by bangerth** [comment]
Like this last commit?

**Pull request merged**